llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -263,9 +254,71 @@ bool RepoPruning::runOnModule(Module &M) {
     }
   }
 
-  LLVM_DEBUG(dbgs() << "size of module: " << M.size() << '\n');
-  LLVM_DEBUG(dbgs() << "size of removed functions: " << NumFunctions << '\n');
-  LLVM_DEBUG(dbgs() << "size of removed variables: " << NumVariables << '\n');
+  LLVM_DEBUG(dbgs() << "Size of module: " << M.size() << '\n');
+  LLVM_DEBUG(dbgs() << "Number of removed functions: " << NumFunctions << '\n');
+  LLVM_DEBUG(dbgs() << "Number of removed variables: " << NumVariables << '\n');
+
+  return Changed;
+}
+
+static bool wasPruned(const GlobalObject &GO) {
+  if (const MDNode *const T = GO.getMetadata(LLVMContext::MD_repo_ticket)) {
+    if (const TicketNode *const MD = dyn_cast<TicketNode>(T)) {
+      return MD->getPruned();
+    }
+  }
+  return false;
+}
+
+static bool eliminateUnreferencedAndPrunedGOs(Module &M) {
+  bool Changed = false;
+
+  // Drop the bodies of available externally functions.
+  for (Function &F : M) {
+    if (F.isDeclaration() || !F.use_empty() || !wasPruned(F))
+      continue;
+    // This will set the linkage to external
+    F.deleteBody();
+    F.removeDeadConstantUsers();
+    NumFunctions++;
+    Changed = true;
+  }
+
+  // Drop initializers of available externally global variables.
+  for (GlobalVariable &GV : M.globals()) {
+    if (GV.isDeclaration() || !GV.use_empty() || !wasPruned(GV))
+      continue;
+    if (GV.hasInitializer()) {
+      Constant *Init = GV.getInitializer();
+      GV.setInitializer(nullptr);
+      if (isSafeToDestroyConstant(Init))
+        Init->destroyConstant();
+    }
+    GV.removeDeadConstantUsers();
+    GV.setLinkage(GlobalValue::ExternalLinkage);
+    NumVariables++;
+    Changed = true;
+  }
+
+  return Changed;
+}
+
+bool RepoPruning::runOnModule(Module &M) {
+  if (skipModule(M))
+    return false;
+
+  bool Changed = doPruning(M);
+
+  // If a function or a global variable which is pruned (already in the
+  // database) and has no live ("un-pruned") callers can have its IR deleted
+  // rather than being marked available_extern since we know that nothing will
+  // attempt to use it for code generation
+  bool IsModified = false;
```

Comment:
IsModified sounds essentially the same as Changed. Is there a better name for IsModified? Also, IsModified could be inside the if  (Changed) to make its scope clearer.

Reply:
Thanks, fixed these:
1) Renamed `Changed` to `IsPruned`;
2) Renamed `IsModified` to `IsEliminated`;
3) Move `IsEliminated` inside ```if (IsPruned)```;